File: iOS/PanoAudioChat/PanoAudioChatSDK/Source/Sections/AudioRoom/View/Player/PanoAudioPanelView.m
//
//  PanoAudioPanelView.m
//  PanoAudioChatSDK
//
//  Created by pano on 2021/7/20.
//

#import "PanoAudioPanelView.h"
#import "PanoAudioVolume.h"
#import "PanoDefine.h"
#import "PanoClientService.h"

@implementation PanoAudioPanelView
{
    PanoAudioVolume *musicVolumeView;
    
    PanoAudioVolume *effectVolumeView;
    
    UIView *contentView;
    
    UIView *coverView;
    
    UILabel *musicTitleLabel;
    
    UILabel *effectTitleLabel;
    
    UIView *lineView;
    
    UIStackView *musicView;
    
    UIStackView *effectAudioView;
    
    UIStackView *effectAudioView2;
    
    NSMutableArray<NSString *> *dataSource;
    
    __weak PanoPlayerService *playerService;
}

- (void)initViews  {
    
    playerService = PanoClientService.service.playerService;
    
    coverView = [UIView new];
    coverView.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.3];
    [coverView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(dismiss)]];
    [self addSubview:coverView];
    
    contentView = [UIView new];
    contentView.backgroundColor = [UIColor whiteColor];
    contentView.layer.cornerRadius = 10;
    contentView.layer.masksToBounds = true;
    [self addSubview:contentView];
    
    musicTitleLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, PanoAppWidth, 50)];
    musicTitleLabel.text = NSLocalizedString(@"背景音乐", nil);
    musicTitleLabel.textColor = defaultTextColor();
    musicTitleLabel.textAlignment = NSTextAlignmentCenter;
    musicTitleLabel.font = fontLarger();
    [contentView addSubview:musicTitleLabel];
    
    musicView = [UIStackView new];
    musicView.axis = UILayoutConstraintAxisHorizontal;
    musicView.distribution = UIStackViewDistributionFillEqually;
    musicView.spacing = 30;
    [contentView addSubview:musicView];
    
    NSArray *audios = playerService.audioNames;
    for (NSInteger i=0; i<audios.count; i++) {
        NSString *title = audios[i] ?: @"";
        UIButton *button = [self createMusicButton: NSLocalizedString(title, nil) image:nil];
        button.tag = i;
        [button addTarget:self action:@selector(chooseMusicAction:) forControlEvents:UIControlEventTouchUpInside];
        [musicView addArrangedSubview:button];
    }
    
    musicVolumeView = [[PanoAudioVolume alloc] init];
    musicVolumeView->slider.maximumValue = 200;
    musicVolumeView->slider.value = 100;
    [musicVolumeView->slider addTarget:self action:@selector(musicVolumeChanged:) forControlEvents:UIControlEventValueChanged];
    [contentView addSubview:musicVolumeView];
    
    lineView = [UIView new];
    lineView.backgroundColor = defaultBorderColor();
    [contentView addSubview:lineView];
    
    effectTitleLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, PanoAppWidth, 50)];
    effectTitleLabel.text = NSLocalizedString(@"声效", nil);
    effectTitleLabel.textColor = defaultTextColor();
    effectTitleLabel.textAlignment = NSTextAlignmentCenter;
    effectTitleLabel.font = fontLarger();
    [contentView addSubview:effectTitleLabel];
    
    effectAudioView = [UIStackView new];
    effectAudioView.axis = UILayoutConstraintAxisHorizontal;
    effectAudioView.distribution = UIStackViewDistributionFillEqually;
    effectAudioView.spacing = 30;
    [contentView addSubview:effectAudioView];
    
    effectAudioView2 = [UIStackView new];
    effectAudioView2.axis = UILayoutConstraintAxisHorizontal;
    effectAudioView2.distribution = UIStackViewDistributionFillEqually;
    effectAudioView2.spacing = 30;
    [contentView addSubview:effectAudioView2];
    NSArray *sounds = playerService.soundNames;
    for (NSInteger i=0; i<3; i++) {
        NSString *title = sounds[i] ?: @"";
        UIButton *button = [self createMusicButton: NSLocalizedString(title, nil) image:nil];
        button.tag = i;
        [button addTarget:self action:@selector(chooseSoundEffectAction:) forControlEvents:UIControlEventTouchUpInside];
        [effectAudioView addArrangedSubview:button];
    }
    for (NSInteger j=3; j<sounds.count; j++) {
        NSString *title = sounds[j] ?: @"";
        UIButton *button = [self createMusicButton: NSLocalizedString(title, nil) image:nil];
        [button addTarget:self action:@selector(chooseSoundEffectAction:) forControlEvents:UIControlEventTouchUpInside];
        button.tag = j;
        [effectAudioView2 addArrangedSubview:button];
    }
    
    
    effectVolumeView = [[PanoAudioVolume alloc] init];
    [effectVolumeView->slider addTarget:self action:@selector(soundEffectVolumeChanged:) forControlEvents:UIControlEventValueChanged];
    effectVolumeView->slider.value = 120;
    [contentView addSubview:effectVolumeView];
}

- (UIButton *)createMusicButton:(NSString *)title image:(UIImage *)image {
    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    [button setTitle:title forState: UIControlStateNormal];
    [button setImage:image forState: UIControlStateNormal];
    [button setImage:image forState: UIControlStateHighlighted];
    [button setTitleColor: defaultTextColor() forState:UIControlStateNormal];
    [button setTitleColor: [UIColor whiteColor] forState:UIControlStateHighlighted];
    [button setTitleColor: [UIColor whiteColor] forState:UIControlStateSelected];
    [button setBackgroundImage:[UIImage pano_imageWithColor:UIColor.whiteColor] forState: UIControlStateNormal];
    [button setBackgroundImage:[UIImage imageNamed:@"home_bg_blue"] forState: UIControlStateSelected];
    [button setBackgroundImage:[UIImage imageNamed:@"home_bg_blue"] forState: UIControlStateHighlighted];
    button.layer.cornerRadius = 8;
    button.layer.masksToBounds = true;
    button.layer.borderColor = defaultBorderColor().CGColor;
    button.layer.borderWidth = 0.8;
    return button;
}

- (void)chooseMusicAction:(UIButton *)sender {
    NSString *title = playerService.audioNames[sender.tag];
    if (title) {
        if ([playerService.activeAudioName isEqualToString:title]) {
            [playerService stopAudioMixingTask:title];
        } else {
            [playerService startAudioMixingTask:title withVolume:(UInt32)musicVolumeView->slider.value];
        }
        [self updateMusicButtonStatus];
    }
}

- (void)chooseSoundEffectAction:(UIButton *)sender {
    NSString *title = playerService.soundNames[sender.tag];
    if (title) {
        [playerService startSoundEffectTask:title withVolume:(UInt32)effectVolumeView->slider.value];
    }
}

- (void)initConstraints  {
    [coverView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.left.right.mas_equalTo(0);
        make.bottom.mas_equalTo(contentView.mas_top);
    }];
    [contentView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.mas_equalTo(0);
        make.height.mas_equalTo(420);
        make.bottom.mas_equalTo(0).offset(5);
    }];
    
    [musicTitleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.left.right.mas_equalTo(0);
        make.height.mas_equalTo(60);
    }];
    
    UIEdgeInsets insets = UIEdgeInsetsMake( 0, 30, 0, 30);
    [musicView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.mas_equalTo(0).insets(insets);
        make.top.mas_equalTo(musicTitleLabel.mas_bottom).offset(10);
        make.height.mas_equalTo(40);
    }];
    
    [musicVolumeView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.mas_equalTo(0);
        make.top.mas_equalTo(musicView.mas_bottom).offset(10);
        make.height.mas_equalTo(50);
    }];
    
    [lineView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.mas_equalTo(0).insets(UIEdgeInsetsMake(0, defaultMargin,0, defaultMargin));
        make.top.mas_equalTo(musicVolumeView.mas_bottom);
        make.height.mas_equalTo(0.8);
    }];
    
    [effectTitleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.mas_equalTo(0);
        make.top.mas_equalTo(lineView.mas_bottom);
        make.height.mas_equalTo(60);
    }];
    
    [effectAudioView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.mas_equalTo(0).insets(insets);
        make.top.mas_equalTo(effectTitleLabel.mas_bottom).offset(10);
        make.height.mas_equalTo(40);
    }];
    
    [effectAudioView2 mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.mas_equalTo(0).insets(insets);
        make.top.mas_equalTo(effectAudioView.mas_bottom).offset(10);
        make.height.mas_equalTo(40);
    }];
    
    [effectVolumeView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.mas_equalTo(0);
        make.top.mas_equalTo(effectAudioView2.mas_bottom).offset(10);
        make.height.mas_equalTo(50);
    }];
}

- (void)showInView:(UIView *)parentView {
    if (![parentView.subviews containsObject:self]) {
        [parentView addSubview:self];
    }
    self.alpha = 0.1;
    CGRect frame = parentView.frame;
    frame.origin.y = frame.size.height;
    self.frame = frame;
    [UIView animateWithDuration:0.25 animations:^{
        self.alpha = 1;
        [self mas_remakeConstraints:^(MASConstraintMaker *make) {
            make.edges.mas_equalTo(0);
        }];
    }];
    [self updateMusicButtonStatus];
}

- (void)updateMusicButtonStatus {
    NSUInteger index = [playerService.audioNames indexOfObject:playerService.activeAudioName];
    for (NSInteger i=0; i<musicView.subviews.count; i++) {
        UIButton *view = musicView.subviews[i];
        view.selected = i == index;
    }
}

- (void)dismiss  {
    [UIView animateWithDuration:0.25 animations:^{
        self.alpha = 0.1;
        self.frame = CGRectMake(0, PanoAppHeight,  PanoAppWidth, PanoAppHeight);
        [self removeFromSuperview];
        if (self->dismissBlock != nil) {
            self->dismissBlock();
        }
    }];
}

- (void)musicVolumeChanged:(UISlider *)slider {
    [playerService updateVolume:(UInt32)slider.value];
}

- (void)soundEffectVolumeChanged:(UISlider *)slider {
    
}

@end

File: iOS/PanoAudioChat/PanoAudioChat/Classes/Sections/AudioRoom/View/Player/PanoAudioPanelView.swift
//
//  PanoAudioPanelView.swift
//  PanoAudioChat
//
//  Created by wenhua.zhang on 2020/11/4.
//

import UIKit

class PanoAudioMusicCell: UITableViewCell {
    
    var audioName: UILabel!
    var audioStateLabel: UILabel!
    var wrapView: UIView!
    
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        
        wrapView = UIView()
        wrapView.layer.cornerRadius = 5
        wrapView.layer.masksToBounds = true
        self.contentView.addSubview(wrapView)
        
        audioName = UILabel()
        audioName.font = FontLittle
        audioName.textColor = DefaultTextColor
        audioName.text = "音乐1"
        wrapView.addSubview(audioName)
        
        audioStateLabel = UILabel()
        audioStateLabel.font = FontLittle
        audioStateLabel.textColor = DefaultTextColor
        audioStateLabel.text = "已暂停"
        wrapView.addSubview(audioStateLabel)
        
        wrapView.snp.makeConstraints { (make) in
            make.edges.equalToSuperview().inset(UIEdgeInsets(top: 0, left: defaultMargin, bottom: 0, right: defaultMargin))
        }
        
        audioName.snp.makeConstraints { (make) in
            make.top.bottom.equalToSuperview()
            make.left.equalToSuperview().offset(8)
        }
        
        audioStateLabel.snp.makeConstraints { (make) in
            make.top.bottom.equalToSuperview()
            make.left.equalTo(audioName.snp.right).offset(defaultMargin)
        }
        
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

class PanoAudioPanelView: PanoBaseView {
    
    var musicVolumeView: PanoAudioVolume!
    
    var effectVolumeView: PanoAudioVolume!
    
    var contentView: UIView!
    
    var coverView: UIView!
    
    var musicTitleLabel: UILabel!
    
    var effectTitleLabel: UILabel!
    
    var lineView: UIView!
    
    var musicView: UIStackView!
    
    var effectAudioView: UIStackView!
    
    var effectAudioView2: UIStackView!
    
    var dataSource = [String]()
    
    var dismissBlock : (() -> ())? = nil
    
    override func initViews() {
        
        coverView = UIView()
        coverView.backgroundColor = UIColor.black.withAlphaComponent(0.3)
        coverView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(dismiss)))
        self.addSubview(coverView)
        
        contentView = UIView()
        contentView.backgroundColor = UIColor.white
        contentView.layer.cornerRadius = 10;
        contentView.layer.masksToBounds = true
        self.addSubview(contentView)
        
        musicTitleLabel = UILabel(frame: CGRect(x: 0, y: 0, width: PanoAppWidth, height: 50))
        musicTitleLabel.text = "背景音乐".localized()
        musicTitleLabel.textColor = DefaultTextColor
        musicTitleLabel.textAlignment = .center
        musicTitleLabel.font = FontLarger
        contentView.addSubview(musicTitleLabel)
        
        musicView = UIStackView()
        musicView.axis = .horizontal
        musicView.distribution = .fillEqually
        musicView.spacing = 30;
        contentView.addSubview(musicView)
        
        
        let audios = PanoPlayerService.service()?.audioNames;
        if let musicCount = audios?.count {
            for index in 0..<musicCount {
                let title = audios?[index] ?? ""
                let button = createMusicButton(title: title.localized(), image: nil)
                button.addTarget(self, action: #selector(chooseMusicAction(sender:)), for: .touchUpInside)
                musicView.addArrangedSubview(button)
            }
        }
        
        musicVolumeView = PanoAudioVolume()
        musicVolumeView.slider.maximumValue = 200;
        musicVolumeView.slider.value = 100;
        musicVolumeView.slider.addTarget(self, action: #selector(musicVolumeChanged(slider:)), for: .valueChanged)
        contentView.addSubview(musicVolumeView)
        
        lineView = UIView()
        lineView.backgroundColor = DefaultBorderColor
        contentView.addSubview(lineView)
        
        
        effectTitleLabel = UILabel(frame: CGRect(x: 0, y: 0, width: PanoAppWidth, height: 50))
        effectTitleLabel.text = "声效".localized()
        effectTitleLabel.textColor = DefaultTextColor
        effectTitleLabel.textAlignment = .center
        effectTitleLabel.font = FontLarger
        contentView.addSubview(effectTitleLabel)
        
        effectAudioView = UIStackView()
        effectAudioView.axis = .horizontal
        effectAudioView.distribution = .fillEqually
        effectAudioView.spacing = 30;
        contentView.addSubview(effectAudioView)
        
        effectAudioView2 = UIStackView()
        effectAudioView2.axis = .horizontal
        effectAudioView2.distribution = .fillEqually
        effectAudioView2.spacing = 30;
        contentView.addSubview(effectAudioView2)
        
        let sounds = PanoPlayerService.service()?.soundNames;
        if let soundCount = sounds?.count {
            for index in 0..<3 {
                let title = sounds?[index] ?? ""
                let button = createMusicButton(title: title.localized(), image: nil)
                button.addTarget(self, action: #selector(chooseSoundEffectAction(sender:)), for: .touchUpInside)
                effectAudioView.addArrangedSubview(button)
            }
            
            for index in 3..<soundCount {
                let title = sounds?[index] ?? ""
                let button = createMusicButton(title: title.localized(), image: nil)
                button.addTarget(self, action: #selector(chooseSoundEffectAction(sender:)), for: .touchUpInside)
                effectAudioView2.addArrangedSubview(button)
            }
        }
        
        effectVolumeView = PanoAudioVolume()
        effectVolumeView.slider.addTarget(self, action: #selector(soundEffectVolumeChanged(slider:)), for: .valueChanged)
        effectVolumeView.slider.value = 120
        contentView.addSubview(effectVolumeView)
    }
    
    func createMusicButton(title: String, image: UIImage?) -> UIButton {
        let button = UIButton(type: .custom)
        button.setTitle(title, for: .normal);
        button.setImage(image, for: .normal);
        button.setImage(image, for: .highlighted)
        button.setTitleColor(DefaultTextColor, for: .normal)
        button.setTitleColor(UIColor.white, for: .highlighted)
        button.setTitleColor(UIColor.white, for: .selected)
        button.setBackgroundImage(UIImage(color: UIColor.white), for: .normal)
        button.setBackgroundImage(UIImage(named: "home_bg_blue"), for: .selected)
        button.setBackgroundImage(UIImage(named: "home_bg_blue"), for: .highlighted)
        button.layer.cornerRadius = 8
        button.layer.masksToBounds = true
        button.layer.borderColor = DefaultBorderColor.cgColor
        button.layer.borderWidth = 0.8;
        return button
    }
    
    @objc func chooseMusicAction(sender: UIButton) {
        guard let title = sender.titleLabel?.text?.localized() else { return  }
        let volume = musicVolumeView.slider.value
        if PanoPlayerService.service()?.activeAudioName == title {
            PanoPlayerService.service()?.stopAudioMixingTask(fileName: title)
        } else {
            PanoPlayerService.service()?.startAudioMixingTask(fileName: title, volume: Int32(volume))
        }
        self.updateStatus(title: PanoPlayerService.service()?.activeAudioName ?? "");
    }
    
    @objc func chooseSoundEffectAction(sender: UIButton) {
        guard let title = sender.titleLabel?.text?.localized() else { return  }
        let volume = effectVolumeView.slider.value
        PanoPlayerService.service()?.startSoundEffectTask(fileName: title, volume: Int32(volume))
    }
    
    override func initConstraints() {
        
        coverView.snp.makeConstraints { (make) in
            make.top.left.right.equalToSuperview()
            make.bottom.equalTo(contentView.snp.top)
        }
        contentView.snp.makeConstraints { (make) in
            make.left.right.equalToSuperview()
            make.height.equalTo(420)
            make.bottom.equalToSuperview().offset(5)
        }
        
        musicTitleLabel.snp.makeConstraints { (make) in
            make.top.left.right.equalToSuperview()
            make.height.equalTo(60)
        }
        
        let insets = UIEdgeInsets(top: 0, left: 30, bottom: 0, right: 30)
        musicView.snp.makeConstraints { (make) in
            make.left.right.equalToSuperview().inset(insets)
            make.top.equalTo(musicTitleLabel.snp.bottom).offset(10)
            make.height.equalTo(40)
        }
      
        musicVolumeView.snp.makeConstraints { (make) in
            make.left.right.equalToSuperview()
            make.top.equalTo(musicView.snp.bottom).offset(10)
            make.height.equalTo(50)
        }
        
        lineView.snp.makeConstraints { (make) in
            make.left.right.equalToSuperview().inset(UIEdgeInsets(top: 0, left: defaultMargin, bottom: 0, right: defaultMargin))
            make.top.equalTo(musicVolumeView.snp.bottom)
            make.height.equalTo(0.8)
        }
        
        effectTitleLabel.snp.makeConstraints { (make) in
            make.left.right.equalToSuperview()
            make.top.equalTo(lineView.snp.bottom)
            make.height.equalTo(60)
        }
        
        effectAudioView.snp.makeConstraints { (make) in
            make.left.right.equalToSuperview().inset(insets)
            make.top.equalTo(effectTitleLabel.snp.bottom).offset(10)
            make.height.equalTo(40)
        }
        
        effectAudioView2.snp.makeConstraints { (make) in
            make.left.right.equalToSuperview().inset(insets)
            make.top.equalTo(effectAudioView.snp.bottom).offset(10)
            make.height.equalTo(40)
        }
        
        effectVolumeView.snp.makeConstraints { (make) in
            make.left.right.equalToSuperview()
            make.top.equalTo(effectAudioView2.snp.bottom).offset(10)
            make.height.equalTo(50)
        }
    }
    
    func showInView(parentView: UIView) {
        if !(parentView.subviews.contains(self)) {
            parentView.addSubview(self)
        }
        self.alpha = 0.1;
        var frame = parentView.frame
        frame.origin.y = frame.size.height
        self.frame = frame
        UIView.animate(withDuration: 0.25) {
            self.alpha = 1
            self.snp.remakeConstraints { (make) in
                make.edges.equalToSuperview()
            }
        }
        self.updateStatus(title: PanoPlayerService.service()?.activeAudioName ?? "")
    }
    
    func updateStatus(title: String) {
        for index in 0..<musicView.subviews.count {
            let view = musicView.subviews[index] as! UIButton
            if view.titleLabel?.text?.localized() == title {
                view.isSelected = true
            } else {
                view.isSelected = false
            }
        }
    }
    
    @objc func dismiss() {
        UIView.animate(withDuration: 0.25) {
            self.alpha = 0.1;
            self.frame = CGRect(x: 0, y: PanoAppHeight, width: PanoAppWidth, height: PanoAppHeight)
            self.removeFromSuperview()
            if (self.dismissBlock != nil) {
                self.dismissBlock!()
            }
        }
    }
    
    @objc func musicVolumeChanged(slider: UISlider) {
        PanoPlayerService.service()?.updateVolume(volume: Int32(slider.value))
    }
    
    @objc func soundEffectVolumeChanged(slider: UISlider) {
        
    }
}
